### PR TITLE
Use pointer for PSP allow escalation

### LINKED ISSUE #53437: unable to deploy privileged pod after 1.8 upgrade unless I set allowPrivilegeEscalation true
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

I upgraded a cluster to 1.8 and ran into an issue with pods that use `privileged: true` and don't set `allowPrivilegeEscalation: true`. 

The PSP I had created and been using prior to 1.8:
```
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: privileged
spec:
  privileged: true
  hostNetwork: true
  hostPID: true
  hostIPC: true
  hostPorts:
  - min: 1
    max: 32000
  fsGroup:
    rule: RunAsAny
  runAsUser:
    rule: RunAsAny
  seLinux:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  volumes:
  - '*'
```

The error I got when applying an update to the canal daemonset:

```
  Warning  FailedCreate  10m                 daemon-set  Error creating: Pod "canal-z8dmj" is invalid: [spec.containers[0].securityContext: Invalid value: api.SecurityContext{Capabilities:(*api.Capabilities)(nil), Privileged:(*bool)(0xc424f9c2c0), SELinuxOptions:(*api.SELinuxOptions)(nil), RunAsUser:(*int64)(nil), RunAsNonRoot:(*bool)(nil), ReadOnlyRootFilesystem:(*bool)(nil), AllowPrivilegeEscalation:(*bool)(0xc427e10c38)}: cannot set `allowPrivilegeEscalation` to false and `privileged` to true, spec.containers[2].securityContext: Invalid value: api.SecurityContext{Capabilities:(*api.Capabilities)(nil), Privileged:(*bool)(0xc424f9c2ca), SELinuxOptions:(*api.SELinuxOptions)(nil), RunAsUser:(*int64)(nil), RunAsNonRoot:(*bool)(nil), ReadOnlyRootFilesystem:(*bool)(nil), AllowPrivilegeEscalation:(*bool)(0xc427e10c38)}: cannot set `allowPrivilegeEscalation` to false and `privileged` to true]
```

**What you expected to happen**:

I expect the pod to be valid since it was not noted as a breaking change in the release notes for 1.8.

**How to reproduce it (as minimally and precisely as possible)**:

Using the PSP above, attempt to create a pod with `privileged: true` and no value for `allowPrivilegeEscalation`. 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.0", GitCommit:"6e937839ac04a38cac63e6a7a306c5d035fe7b0a", GitTreeState:"clean", BuildDate:"2017-09-28T22:57:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.0", GitCommit:"6e937839ac04a38cac63e6a7a306c5d035fe7b0a", GitTreeState:"clean", BuildDate:"2017-09-28T22:46:41Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
```

cc @liggitt @tallclair @jessfraz

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #53437 

The `AllowPrivilegeEscalation` field was added to PodSpec and PodSecurityPolicySpec in 1.8.0.

In order to remain compatible with pre-1.8.0 behavior, PodSecurityPolicy objects created against a previous release must not restrict this field, which means the field must default to true in PodSecurityPolicySpec. However, the field was added as a `bool`, not a `*bool`, which means that no defaulting is possible.

We have two options:
1. Require all pre-existing PodSecurityPolicy objects that intend to allow privileged permissions to update to set this new field to true
2. Change the field to a `*bool` and default it to true.

This PR does the latter. With this change, we have the following behavior:

A 1.8.1+ client/server now has three ways to serialize:
* `nil` values are dropped from serialization (because `omitempty`), which is interpreted correctly by other 1.8.1+ clients/servers, and is interpreted as false by 1.8.0
* `false` values are serialized and interpreted correctly by all clients/servers
* `true` values are serialized and interpreted correctly by all clients/servers

A 1.8.0 client/server has two ways to serialize:
* `false` values are dropped from serialization (because `omitempty`), which is interpreted as `false` by other 1.8.0 clients/servers, but as `nil` (and therefore defaulting to true) by 1.8.1+ clients/servers
* `true` values are serialized and interpreted correctly by all clients/servers

The primary concern is the 1.8.0 server dropping the `false` value from serialization, but I consider the compatibility break with pre-1.8 behavior to be more severe, especially if we can resolve the regression in an immediate point release.

```release-note
PodSecurityPolicy: Fixes a compatibility issue that caused policies that previously allowed privileged pods to start forbidding them, due to an incorrect default value for `allowPrivilegeEscalation`. PodSecurityPolicy objects defined using a 1.8.0 client or server that intended to set `allowPrivilegeEscalation` to `false` must be reapplied after upgrading to 1.8.1.
```

### CODE REVIEW

File: staging/src/k8s.io/api/extensions/v1beta1/types.go [hunk 1]
```diff
@@ -975,9 +975,9 @@ type PodSecurityPolicySpec struct {
 	// +optional
 	DefaultAllowPrivilegeEscalation *bool `json:"defaultAllowPrivilegeEscalation,omitempty" protobuf:"varint,15,opt,name=defaultAllowPrivilegeEscalation"`
 	// AllowPrivilegeEscalation determines if a pod can request to allow
-	// privilege escalation.
+	// privilege escalation. If unspecified, defaults to true.
```
Comment (sttts) [inline staging/src/k8s.io/api/extensions/v1beta1/types.go]: Is this an API change? undefined was leading to `false` before.

Moreover, I expected to be safe by default, i.e. default to false.

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-api-reviews @kubernetes/api-reviewers

**Comment by liggitt** [comment]
@jessfraz @tallclair PTAL

File: staging/src/k8s.io/api/extensions/v1beta1/types.go (hunk 1 above)
Comment (tallclair) [reply staging/src/k8s.io/api/extensions/v1beta1/types.go]: Prior to 1.8, `AllowPrivilegeEscalation` was implicitly always true. But I think this needs to be true to fix https://github.com/kubernetes/kubernetes/issues/53437, which motivated this change.

Comment (liggitt) [reply staging/src/k8s.io/api/extensions/v1beta1/types.go]: > Is this an API change? undefined was leading to false before.

Correct, which was a breaking change in 1.8.0

> Moreover, I expected to be safe by default, i.e. default to false.

Compatibility wins. The ability to restrict this did not exist prior to 1.8.0, so policies that do not include this field must continue pre-1.8.0 behavior.

File: pkg/apis/extensions/types.go [hunk 2]
```diff
@@ -937,9 +937,9 @@ type PodSecurityPolicySpec struct {
 	// +optional
 	DefaultAllowPrivilegeEscalation *bool
 	// AllowPrivilegeEscalation determines if a pod can request to allow
-	// privilege escalation.
+	// privilege escalation. If unspecified, defaults to true.
 	// +optional
-	AllowPrivilegeEscalation bool
+	AllowPrivilegeEscalation *bool
```
Comment (mikedanese) [inline pkg/apis/extensions/types.go]: Can you leave it bool on the internal type?

**Review by tallclair — COMMENTED** [review]
Thanks for fixing this Jordan. How does an API change like this work mid-release? Is it OK since the it's backwards compatible with the serialized (json & proto) formats?

File: pkg/apis/extensions/types.go (hunk 2 above)
Comment (liggitt) [reply pkg/apis/extensions/types.go]: since we're defaulting... maybe... will try

Comment (mikedanese) [reply pkg/apis/extensions/types.go]: See:

https://github.com/kubernetes/kubernetes/blob/6039c79f6c583084a2959807f516490fe7232840/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/conversion.go#L141-L148

Comment (liggitt) [reply pkg/apis/extensions/types.go]: yeah, updated

**Comment by liggitt** [comment]
> How does an API change like this work mid-release? Is it OK since the it's backwards compatible with the serialized (json & proto) formats?

in 1.8.0, the only value serialized is true (false is omitted)
with this change in 1.8.1+, the only values serialized are true/false (nil is omitted)

* 1.8.0 and 1.8.1+ clients/servers can decode all serialized data
* 1.8.0 treats undefined as `false`, which is the breaking behavior we want to fix
* 1.8.0 drops explicit `false` values on reserialization, which a 1.8.1+ server would then treat as `true`. That's why this PR has a `release-note-action-required` label... someone making use of this new field needs to reset the field to false in 1.8.1

**Comment by smarterclayton** [comment]
API changes approved

/approve

**Comment by jessfraz** [comment]
LGTM thanks for doing this

**Pull request merged**